src/perception/aruco_processing.cpp

```diff
@@ -0,0 +1,136 @@
+ * Later camera space pose information is filled in when we receive point cloud data.
+ *
+ * @param msg
+ */
+void FiducialsNode::imageCallback(sensor_msgs::ImageConstPtr const& msg) {
+    if (!mEnableDetections) return;
+
+    if (mIsVerbose) {
+        ROS_INFO("Got image %d", msg->header.seq);
+    }
+
+    try {
+        mCvPtr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
+
+        // Detect the fiducial vertices in screen space and their respective ids
+        cv::aruco::detectMarkers(mCvPtr->image, mDictionary, mCornersCache, mIdsCache, mDetectorParams);
+
+        // Save fiducials currently on screen to the immediate buffer
+        for (size_t i = 0; i < mIdsCache.size(); ++i) {
+            int id = mIdsCache[i];
+            cv::Point2f imageCenter = std::accumulate(mCornersCache[i].begin(), mCornersCache[i].end(), cv::Point2f{}) / 4.0f;
+            mImmediateFiducials[id].id = id;
+            mImmediateFiducials[id].imageCenter = imageCenter;
+        }
+
+        // Remove fiducials in the immediate buffer that are no longer in sight
+        auto it = mImmediateFiducials.begin();
+        while (it != mImmediateFiducials.end()) {
+            if (std::find(mIdsCache.begin(), mIdsCache.end(), it->first) == mIdsCache.end()) {
+                it = mImmediateFiducials.erase(it);
+            } else {
+                ++it;
+            }
+        }
+
+        fiducial_msgs::FiducialTransformArray fidArray{};
+        fidArray.header.frame_id = ODOM_FRAME;
+        fidArray.header.stamp = ros::Time::now();
+        fidArray.header.seq = mSeqNum;
+        fidArray.transforms.reserve(mPersistentFiducials.size());
+
+        // Add readings to the persistent representations of the fiducials
+        for (auto [id, immediateFid]: mImmediateFiducials) {
+            PersistentFiducial& fid = mPersistentFiducials[id];
+            if (!immediateFid.fidInCam.has_value()) continue; // This is set if the point cloud had a valid reading for this fiducial
+
+            fid.id = id;
+            SE3 fidInOdom = SE3::transform(mTfBuffer, ROVER_FRAME, ODOM_FRAME, immediateFid.fidInCam.value(), mSeqNum);
+            fid.setFilterParams(mFilterCount, mFilterProportion);
+            fid.addReading(fidInOdom);
+        }
+
+        // Send all transforms of persistent fiducials
+        for (auto [id, fid]: mPersistentFiducials) {
+            if (!fid.fidInOdomX.ready()) continue; // Wait until the filters have enough readings to become meaningful
+
+            SE3::sendTransform(mTfBroadcaster, "fiducial" + std::to_string(id), ODOM_FRAME, fid.getPose(), mSeqNum);
+        }
+
+        mFidPub.publish(fidArray);
+
+        size_t detectedCount = mIdsCache.size();
+        if (mIsVerbose || !mPrevDetectedCount.has_value() || detectedCount != mPrevDetectedCount.value()) {
+            mPrevDetectedCount = detectedCount;
+            ROS_INFO("Detected %zu markers", detectedCount);
```

Comment:
please make a note of using the ROS debug macros in the wiki under best practices

Reply:
Done